libethereum/BlockChain.cpp

```diff
@@ -194,30 +194,26 @@ void BlockChain::init(ChainParams const& _p)
 
 bool BlockChain::open(fs::path const& _path, WithExisting _we)
 {
-    auto const path = _path.empty() ? db::databasePath() : _path;
-    auto const chainPath = path / fs::path(toHex(m_genesisHash.ref().cropped(0, 4)));
-    auto const chainSubPathBlocks = chainPath / fs::path("blocks");
-    auto const extrasPath = chainPath / fs::path(toString(c_databaseVersion));
-    auto const extrasSubPathExtras = extrasPath / fs::path("extras");
     unsigned lastMinor = c_databaseMinorVersion;
     bool rebuildNeeded = false;
 
     if (db::isDiskDatabase())
     {
+        if (!_path.empty())
```

Comment:
@gumb0: We can, but that seems unnecessary if m_dbPaths is already set?

Reply:
You could then just 
```
if (!m_dbPaths.pathsSet()) 
  m_dbPaths.setPaths(_path, m_genesisHash);
```

But I would probably make it `unique_ptr` and
```
if (!m_dbPaths) 
  m_dbPaths = std::make_unique<DatabasePaths>(_path, m_genesisHash);
```

Reply:
@gumb0 unique_ptr makes sense, but I like the current behavior of only setting the path if the supplied path arg is empty since I think that's more intuitive, i.e. if I call a function and pass it an arg I expect that arg to be used and not silently dropped. What do you think?

Reply:
Makes sense, but I don't like the hidden features of the function interface, like magic value "empty path" meaning "use previous path".
How about instead of empty path calling it with the current path (as before) and then here:
```
if (!m_dbPaths || m_dbPaths.rootPath() != _path) 
  m_dbPaths = std::make_unique<DatabasePaths>(_path, m_genesisHash);
```

then it's kind of an optimization to not reset the paths when not needed. And in practice re-creation will never happen, because it's called with the same path only.